OSDOCS-1594 - Adding pod identity webhook content for ROSA

pneedle-rh opened the pull request:
This applies to `main`, `enterprise-4.12` and `enterprise-4.11`.

This relates to https://issues.redhat.com/browse/OSDOCS-1594. The PR adds an "Assuming an AWS IAM role for a service account" page to the ROSA documentation.

The preview is at https://51026--docspreview.netlify.app/openshift-rosa/latest/authentication/assuming-an-aws-iam-role-for-a-service-account.html.

michaelryanmcneill commented:
A few pieces of feedback in addition to what we discussed in Slack:
- It would be nice to show the optional regional endpoints and token expiration annotations as part of the service account creation (https://github.com/aws/amazon-eks-pod-identity-webhook/blob/master/README.md, see step #3). While they're not required, I have dealt with multiple customers who specifically were interested in those flags. The EKS documentation has a separate section specifically dedicated to that: https://docs.aws.amazon.com/eks/latest/userguide/configure-sts-endpoint.html, but we don't necessarily have to follow that. 
- It would be nice to have a cross-account IAM permissions section, similar to EKS docs, because many customers ask for how to do this. (https://docs.aws.amazon.com/eks/latest/userguide/cross-account-access.html). Because customer's can't create secondary OIDC providers for their cluster, they would have to follow the chaining approach.

pneedle-rh commented:
@michaelryanmcneill thank you for your feedback, this is very helpful. I am on PTO next week and will add your suggestions in this PR when I return. Thanks again!

pneedle-rh commented:
@michaelryanmcneill after some initial testing, I decided to create separate docs Jiras for your requests to add the optional annotations and the cross-account IAM permissions section. We can address those items later, separately from this PR.

The Jiras are as follows:

* [OSDOCS-4492](https://issues.redhat.com/browse/OSDOCS-4492) - [Feature] Add regional endpoint annotations to the pod identity webhook content
* [OSDOCS-4493](https://issues.redhat.com/browse/OSDOCS-4493) - [Feature] Add a cross-account permissions section to the pod identity webhook content

In my testing, the regional endpoint annotation did not produce the corresponding environment variable in the pod. This might potentially relate to https://github.com/aws/amazon-eks-pod-identity-webhook/issues/123. Additionally, I only have access to one AWS account, so I was not able to test the cross-account permissions functionality.

My view is that we should focus on reviewing and publishing the content that I have created for this initial PR and then those items can be tested and added later. In this PR, I have also added some commented out content in relation to the optional annotations for later use.

jeana-redhat reviewed (commented):
This is a really excellent presentation of a workflow that is not at all trivial to explain. 

Most of my feedback here is pretty minor, and some is just checking for my own understanding since I will need to understand this later :nerd_face: Feel free to ping me on Slack if you want to talk about any of the things I brought up more synchronously.

jianping-shu commented on authentication/assuming-an-aws-iam-role-for-a-service-account.adoc:
```diff
@@ -0,0 +1,43 @@
+{product-title} with AWS Security Token Service (STS) clusters include a pod identity webhook for use with pods running in user-defined projects.
+endif::openshift-rosa[]
+
+You can use the pod identity webhook to enable a service account to automatically assume an AWS Identity and Access Management (IAM) role in your own pods. If the assumed IAM role has the required AWS permissions, the pods can run AWS SDK operations for your AWS account by using temporary STS credentials.
```
Personally I think "run AWS SDK operations for your AWS account" can be just changed to "run AWS SDK operations". There are several places with same issues in the chapter.

pneedle-rh replied:
Thank you, that's a good suggestion. I will remove "for your AWS account" throughout the SDK references in the PR.

jianping-shu commented on modules/understanding-pod-identity-webhook-workflow-in-user-defined-projects.adoc:
```diff
@@ -0,0 +1,66 @@
+The following diagram illustrates the pod identity webhook workflow in user-defined projects:
+
+.Pod identity webhook workflow in user-defined projects
+image::pod-identity-webhook-workflow-in-user-defined-projects.png[Pod identity webhook workflow]
```
In my understanding, in STS mode, there are no more access/secret key on AWS side. Pod uses token (with the specified role on AWS side) to access AWS service. So better to remove these to avoid misleading.
See Figure 1. STS authentication flow at https://docs.openshift.com/container-platform/4.11/authentication/managing_cloud_provider_credentials/cco-mode-sts.html#cco-mode-sts

pneedle-rh replied:
> The image has some text in light grey at the bottom right (`286_OpenShift_1122`) - not sure if that's intended to show up

@jeana-redhat thank you for highlighting that! I will request an update to the diagram.

pneedle-rh replied:
> In my understanding, in STS mode, there are no more access/secret key on AWS side. Pod uses token (with the specified role on AWS side) to access AWS service. So better to remove these to avoid misleading. See Figure 1. STS authentication flow at https://docs.openshift.com/container-platform/4.11/authentication/managing_cloud_provider_credentials/cco-mode-sts.html#cco-mode-sts

@jianping-shu thank you for clarifying that. I will remove the access and secret key references from the module and will request an update to the diagram.

pneedle-rh replied:
The diagram has now been updated accordingly.

jianping-shu commented on modules/deploying-a-pod-that-includes-an-aws-sdk.adoc:
```diff
@@ -0,0 +1,66 @@
+
+* You have created an AWS Identity and Access Management (IAM) role for your service account. For more information, see _Setting up an AWS IAM role for a service account_.
+* You have access to a {product-title} cluster that uses the AWS Security Token Service (STS). Admin-level user privileges are not required.
+* You have installed the OpenShift CLI (`oc`).
```
Does it need one more here You have created the service account?

jianping-shu commented on modules/creating-an-example-aws-sdk-container-image.adoc:
```diff
@@ -0,0 +1,67 @@
+
+:_content-type: PROCEDURE
+[id="creating-an-example-aws-sdk-container-image_{context}"]
+= Creating an example AWS SDK container image
```
I think create role&create service service account is a must for all users. But the sections of creating image/pod highly depend on user case. Maybe a note can be added that this is just a sample to show how to deploy and validate a pod.

pneedle-rh replied:
Thanks for highlighting this. I will update "method" to "example method" in the first line.

jianping-shu commented on modules/creating-a-service-account-in-your-project.adoc:
```diff
@@ -0,0 +1,95 @@
+  name: <service_account_name> <1>
+  namespace: <project_name> <2>
+  annotations:
+    eks.amazonaws.com/role-arn: "<aws_iam_role_arn>" <3>
```
The OIDC of OCP STS cluster looks like this
arn:aws:iam::301721915996:oidc-provider/jshu-1116sts-2911-oidc.s3.us-east-2.amazonaws.com
vs. EKS cluster one
"arn:aws:iam::111122223333:oidc-provider/oidc.REGION.eks.amazonaws.com/CLUSTER_ID"

So for OCP STS cluster, still set annotation to "eks.amazonaws.com/role-arn: "???

pneedle-rh replied:
My understanding is that the pod identity webhook is looking for the `eks.amazonaws.com/role-arn` annotation. 

@sjenning can you please advise? Thanks!

sjenning replied:
https://github.com/openshift/cloud-credential-operator/blame/c34501f3d66f132d21a2a9620a886c4b144e1571/bindata/v4.1.0/aws-pod-identity-webhook/deployment.yaml#L32

Yes, unfortunately we never changed it before we rolled it out GA so now it is stuck that way

For built in OCP components, we do not use the webhook to inject role information since we can't be sure the CCO is running by the time the role is used.

pneedle-rh replied:
Thank you for confirming that. I will subsequently keep the annotation as it is in the code block and will set this comment thread to resolved. Thanks again!

pneedle-rh commented on authentication/assuming-an-aws-iam-role-for-a-service-account.adoc:
```diff
@@ -0,0 +1,43 @@
+
+You can use the pod identity webhook to enable a service account to automatically assume an AWS Identity and Access Management (IAM) role in your own pods. If the assumed IAM role has the required AWS permissions, the pods can run AWS SDK operations for your AWS account by using temporary STS credentials.
+
+You can test this feature by following the procedures in this document in sequence.
```
Thank you for highlighting this. I will remove this line.

pneedle-rh commented on authentication/assuming-an-aws-iam-role-for-a-service-account.adoc:
```diff
@@ -0,0 +1,43 @@
+
+[role="_abstract"]
+ifdef::openshift-rosa[]
+{product-title} with AWS Security Token Service (STS) clusters include a pod identity webhook for use with pods running in user-defined projects.
```
Good suggestion, thanks.

pneedle-rh commented on modules/assuming-an-aws-iam-role-in-your-own-pods.adoc:
```diff
@@ -0,0 +1,17 @@
+
+:_content-type: PROCEDURE
+[id="assuming-an-aws-iam-role-in-your-own-pods_{context}"]
+= Assuming an AWS IAM role in your own pods
```
Thank you for this suggestion. Resolved.

pneedle-rh commented on modules/assuming-an-aws-iam-role-in-your-own-pods.adoc:
```diff
@@ -0,0 +1,17 @@
+[id="assuming-an-aws-iam-role-in-your-own-pods_{context}"]
+= Assuming an AWS IAM role in your own pods
+
+Follow the procedures in this section to enable a service account to assume an AWS Identity and Access Management (IAM) role in a pod deployed in a user-defined project.
+
+By following the steps in the procedures, you can complete these tasks:
```
Resolved.

pneedle-rh commented on modules/assuming-an-aws-iam-role-in-your-own-pods.adoc:
```diff
@@ -0,0 +1,17 @@
+
+* Create the required resources, including an AWS IAM role, a service account, a container image that includes an AWS SDK, and a pod deployed by using the image.
+
+* After the pod deployment, verify that the pod identity webhook mutated the AWS environment variables, the volume mount, and the token volume into your pod.
```
My view is that it is better to use "mutated" here, because it does not mutate after the pod deployment but as part of it.

pneedle-rh commented on modules/creating-a-service-account-in-your-project.adoc:
```diff
@@ -0,0 +1,95 @@
+You are automatically switched to the project when it is created.
+====
+
+. Add the following service account configuration to a file named `test-service-account.yaml`:
```
This is something that they can create based on the example in the docs.

pneedle-rh commented on modules/creating-a-service-account-in-your-project.adoc:
```diff
@@ -0,0 +1,95 @@
+//    eks.amazonaws.com/token-expiration: "86400" <5>
+<1> Replace `<service_account_name>` with the name of your service account. The name must match the service account name that you specified in your AWS IAM role configuration.
+<2> Replace `<project_name>` with the name of your project. The name must match the project name that you specified in your AWS IAM role configuration.
+<3> Specifies the ARN of the AWS IAM role that will be assumed by the service account for use within your pod. Replace `<aws_iam_role_arn>` with the ARN for the AWS IAM role that you created for your service account. The format of the role ARN is `arn:aws:iam::<aws_account_id>:role/<aws_iam_role_name>`.
```
Yes I think so, thanks. Resolved.

pneedle-rh commented on modules/creating-an-example-aws-sdk-container-image.adoc:
```diff
@@ -0,0 +1,67 @@
+
+The steps in this procedure provide a method to create a container image that includes an AWS SDK.
+
+This procedure uses Podman to create the container image and quay.io to host the image. The container image can be used to deploy pods that can run AWS SDK operations for your AWS account.
```
Thanks for highlighting this. Resolved throughout the PR.

pneedle-rh commented on modules/creating-an-example-aws-sdk-container-image.adoc:
```diff
@@ -0,0 +1,67 @@
+
+[NOTE]
+====
+In this example procedure, the AWS Boto3 SDK for Python is installed into a container image. For more information about installing and using the AWS Boto3 SDK, see the link:https://boto3.amazonaws.com/v1/documentation/api/latest/index.html[AWS Boto3 documentation]. For details about other AWS SDKs, see link:https://docs.aws.amazon.com/sdkref/latest/guide/overview.html[AWS SDKs and Tools Reference ...
```
Good suggestion, thanks. Resolved.

pneedle-rh commented on modules/deploying-a-pod-that-includes-an-aws-sdk.adoc:
```diff
@@ -0,0 +1,66 @@
+* You have created an AWS Identity and Access Management (IAM) role for your service account. For more information, see _Setting up an AWS IAM role for a service account_.
+* You have access to a {product-title} cluster that uses the AWS Security Token Service (STS). Admin-level user privileges are not required.
+* You have installed the OpenShift CLI (`oc`).
+* You have a container image that includes the AWS Boto3 SDK for Python and the image is available to your cluster. For detailed steps, see _Creating an example AWS SDK container image_.
```
Great suggestion, thank you. Resolved.

pneedle-rh commented on modules/deploying-a-pod-that-includes-an-aws-sdk.adoc:
```diff
@@ -0,0 +1,66 @@
+
+.Procedure
+
+. Add the following pod configuration to a file named `awsboto3sdk-pod.yaml`:
```
Thanks for suggesting this. Resolved.

pneedle-rh commented on modules/setting-up-an-aws-iam-role-a-service-account.adoc:
```diff
@@ -0,0 +1,85 @@
+.Prerequisites
+
+* You have the permissions required to install and configure IAM roles in your AWS account.
+* You have access to a {product-title} cluster that uses the AWS Security Token Service (STS). Admin-level user privileges are not required.
```
No additional requirements, just a user account, as far as I am aware from my testing. I intentionally tested and documented a method to complete the end-to-end sequence without needing to be an administrator.

pneedle-rh commented on modules/setting-up-an-aws-iam-role-a-service-account.adoc:
```diff
@@ -0,0 +1,85 @@
+
+.Procedure
+
+. Add the following JSON configuration to a file named `trust-policy.json`:
```
Great suggestion, thanks! Resolved.

pneedle-rh commented on modules/setting-up-an-aws-iam-role-a-service-account.adoc:
```diff
@@ -0,0 +1,85 @@
+====
+--
+
+. Create an AWS IAM role that uses the trust policy that is defined in the `trust-policy.json` file:
```
My preference is to leave that out in favour of minimalism. I think that speaks for itself.

pneedle-rh commented on modules/setting-up-an-aws-iam-role-a-service-account.adoc:
```diff
@@ -0,0 +1,85 @@
++
+[source,terminal]
+----
+$ aws iam create-role --role-name <aws_iam_role_name> --assume-role-policy-document file://trust-policy.json <1>
```
Good suggestion, thanks. Resolved.

pneedle-rh commented on modules/setting-up-an-aws-iam-role-a-service-account.adoc:
```diff
@@ -0,0 +1,85 @@
++
+[source,terminal]
+----
+$ aws iam attach-role-policy --policy-arn arn:aws:iam::aws:policy/ReadOnlyAccess --role-name <aws_iam_role_name> <1>
```
Good suggestion, thanks.

pneedle-rh commented on modules/understanding-pod-identity-webhook-workflow-in-user-defined-projects.adoc:
```diff
@@ -0,0 +1,66 @@
+
+The workflow has the following stages:
+
+. Within a user-defined project, a user creates a service account that includes an `eks.amazonaws.com/role-arn` annotation. The annotation points to the ARN of the AWS IAM role that you want your service account to assume.
```
Thank you for highlighting this. I think it would definitely need to be something that is done at a later date, given that I am leaving on Friday.

pneedle-rh commented on modules/understanding-pod-identity-webhook-workflow-in-user-defined-projects.adoc:
```diff
@@ -0,0 +1,66 @@
+
+** An `$AWS_ARN_ROLE` environment variable that contains the ARN for the IAM role that has the permissions required to run AWS SDK operations for your AWS account.
+
+** An `$AWS_WEB_IDENTITY_TOKEN_FILE` environment variable that contains the full path in the pod to the OIDC token for the service account. The full path is `/var/run/secrets/eks.amazonaws.com/serviceaccount/token`.
+
+** An `aws-iam-token` volume mounted on the mount point `/var/run/secrets/eks.amazonaws.com/serviceaccount`. An OIDC token file named `token` is contained in the volume. 
```
My preference is to mention the two environment variables adjacently.

pneedle-rh commented on modules/understanding-pod-identity-webhook-workflow-in-user-defined-projects.adoc:
```diff
@@ -0,0 +1,66 @@
+
+** An `$AWS_WEB_IDENTITY_TOKEN_FILE` environment variable that contains the full path in the pod to the OIDC token for the service account. The full path is `/var/run/secrets/eks.amazonaws.com/serviceaccount/token`.
+
+** An `aws-iam-token` volume mounted on the mount point `/var/run/secrets/eks.amazonaws.com/serviceaccount`. An OIDC token file named `token` is contained in the volume. 
```
Good catch, thank you. Resolved in line 39.

pneedle-rh commented on modules/understanding-pod-identity-webhook-workflow-in-user-defined-projects.adoc:
```diff
@@ -0,0 +1,66 @@
+
+** The identity signature is valid and signed by the private key.
+
+** The `sts.amazonaws.com` audience is listed in the OIDC token and matches the audience configured in the OIDC provider.
```
I would say that the note is informational and provides useful context only. The bullet is informational, but it is also an important aspect required for an understanding of the workflow that this section is providing.

I can't think of any other required notes associated with the workflow items at this time.

pneedle-rh commented on modules/understanding-pod-identity-webhook-workflow-in-user-defined-projects.adoc:
```diff
@@ -0,0 +1,66 @@
+The pod identity webhook applies the `sts.amazonaws.com` audience to the OIDC token by default.
+ifdef::openshift-rosa[]
+
+In {product-title} with STS clusters, the OpenID Connect (OIDC) provider is created during install and set as the service account issuer by default. The `sts.amazonaws.com` audience is set by default in the OIDC provider.
```
Good catch, thanks. Resolved.

pneedle-rh commented on modules/verifying-the-assumed-iam-role-in-your-pod.adoc:
```diff
@@ -0,0 +1,163 @@
+[id="verifying-the-assumed-iam-role-in-your-pod_{context}"]
+= Verifying the assumed IAM role in your pod
+
+After deploying an `awsboto3sdk` pod in your project, verify that the pod has been mutated successfully by the pod identity webhook. Check that the required AWS environment variables, volume mount, and OIDC token volume are present within the pod.
```
Yes I think it does, thanks. Resolved.

pneedle-rh commented on modules/verifying-the-assumed-iam-role-in-your-pod.adoc:
```diff
@@ -0,0 +1,163 @@
+[id="verifying-the-assumed-iam-role-in-your-pod_{context}"]
+= Verifying the assumed IAM role in your pod
+
+After deploying an `awsboto3sdk` pod in your project, verify that the pod has been mutated successfully by the pod identity webhook. Check that the required AWS environment variables, volume mount, and OIDC token volume are present within the pod.
```
Present tense I think works. I have also incorporated your active voice suggestion. Thanks!

pneedle-rh commented on modules/verifying-the-assumed-iam-role-in-your-pod.adoc:
```diff
@@ -0,0 +1,163 @@
+
+After deploying an `awsboto3sdk` pod in your project, verify that the pod has been mutated successfully by the pod identity webhook. Check that the required AWS environment variables, volume mount, and OIDC token volume are present within the pod.
+
+Also, verify that the service account assumes the AWS Identity and Access Management (IAM) role for your AWS account when you run AWS SDK operations in the pod.
```
Resolved. Thanks!

pneedle-rh commented on modules/verifying-the-assumed-iam-role-in-your-pod.adoc:
```diff
@@ -0,0 +1,163 @@
+* You have created an AWS IAM role for your service account. For more information, see _Setting up an AWS IAM role for a service account_.
+* You have access to a {product-title} cluster that uses the AWS Security Token Service (STS). Admin-level user privileges are not required.
+* You have installed the OpenShift CLI (`oc`).
+* You have deployed an `awsboto3sdk` pod in your user-defined project. The pod includes the AWS Boto3 SDK for Python. The pod references the service account that uses the pod identity webhook to assume the AWS IAM role required to run the AWS SDK operations. For detailed steps, see _Deploying a pod that includes an AWS SDK_.
```
Good suggestion, thanks. Resolved.

pneedle-rh commented on modules/verifying-the-assumed-iam-role-in-your-pod.adoc:
```diff
@@ -0,0 +1,163 @@
+    TokenExpirationSeconds:  86400
+...
+----
+<1> Lists the `AWS_ROLE_ARN` environment variable that was injected into the pod by the pod identity webhook. The variable contains the Amanzon Resource Name (ARN) of the AWS IAM role to be assumed by the service account.
```
Good catch, thanks! Resolved.

pneedle-rh commented on modules/verifying-the-assumed-iam-role-in-your-pod.adoc:
```diff
@@ -0,0 +1,163 @@
+----
+arn:aws:iam::<aws_account_id>:role/<aws_iam_role_name> <1>
+----
+<1> The output must specify the Amazon Resource Name (ARN) for the AWS IAM role that has the permissions required to run AWS SDK operations for your AWS account.
```
Thanks for highlighting this. Resolved.

pneedle-rh commented on modules/understanding-pod-identity-webhook-workflow-in-user-defined-projects.adoc:
```diff
@@ -0,0 +1,66 @@
+
+. If the project and service account are in the scope of the trust policy for the IAM role that is being assumed, then authorization succeeds.
+
+. After successful authentication and authorization, temporary AWS STS credentials in the form of an access key, a secret key, and a session token are passed to the pod for use by the service account. By using the credentials, the service account is temporarily granted the AWS permissions enabled in the IAM role.
```
Good catch, thank you. The diagram should also say "Session token". I will request an update to the diagram.

pneedle-rh replied:
The diagram has now been updated accordingly.

michaelryanmcneill commented:
This looks great to me, thanks @pneedle-rh for all your hard work.

pneedle-rh commented:
/cherrypick enterprise-4.12

pneedle-rh commented:
/cherrypick enterprise-4.11
